support/cas-server-support-inwebo-mfa/src/main/java/org/apereo/cas/support/inwebo/authentication/InweboAuthenticationHandler.java

```diff
@@ -0,0 +1,49 @@
+package org.apereo.cas.support.inwebo.authentication;
+
+import org.apereo.cas.authentication.AuthenticationHandlerExecutionResult;
+import org.apereo.cas.authentication.Credential;
+import org.apereo.cas.authentication.handler.support.AbstractPreAndPostProcessingAuthenticationHandler;
+import org.apereo.cas.authentication.principal.PrincipalFactory;
+import org.apereo.cas.configuration.model.support.mfa.InweboMultifactorProperties;
+import org.apereo.cas.services.ServicesManager;
+
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+
+/**
+ * The Inwebo authentication handler.
+ *
+ * @author Jerome LELEU
+ * @since 6.3.0
+ */
+@Slf4j
+public class InweboAuthenticationHandler extends AbstractPreAndPostProcessingAuthenticationHandler {
+
+    public InweboAuthenticationHandler(final ServicesManager servicesManager,
+                                       final PrincipalFactory principalFactory,
+                                       final InweboMultifactorProperties inweboProperties) {
+        super(inweboProperties.getName(),
+              servicesManager,
+              principalFactory,
+              inweboProperties.getOrder());
+    }
+
+    @Override
+    protected AuthenticationHandlerExecutionResult doAuthentication(final Credential credential) {
+        val login = credential.getId();
+        LOGGER.info("Creating principal result for: {}", login);
+        val pushCredential = (InweboCredential) credential;
+        val principal = this.principalFactory.createPrincipal(login);
+        return createHandlerResult(pushCredential, principal);
+    }
```

Comment:
This doesn't seem right. You are not actually doing authentication, when the method asks for doAuthentication. The credential needs to be verified somehow. If you're doing that elsewhere, you'd need to move that logic into the handler here.

Reply:
I had that in mind. This is done on purpose. Let me explain.

In fact, for the push notification, we need to call Inwebo to check if the user has accepted the notification on his mobile/desktop app. The response can be "yes" or "timeout" if the user has waited too long or "refused" if the user has refused. But this can also be "wait" if the user has not accepted yet and there remains time. And in that case, we must come back on the "waiting screen".
This is why I control this behavior in a webflow action: https://github.com/apereo/cas/pull/5001/files#diff-f8b3de0873b2467c5a3d782d751949279eec743827a82a353b33ee9795cdbaacR59

For the browser authentication, I reused the same logic to only have one `InweboCredential` and `InweboAuthenticationHandler` for both authentication methods (mobile/desktop + browser).

Does it make sense?